### ISSUE TITLE
Modals are slow to open

### ISSUE THREAD

**Issue by finger563** [Max-Mobility/permobil-client#749]
esp. the first time they load

**Comment by bradmartin** [comment]
we have a `performance.now` function that prints nanotime to console from the `@permobil/nativescript` module that you can use to wrap around diff functions and see where slow down might be. I'll probably start looking at the constructor of the VMs for the modals. We might benefit just wrapping them in a simple timeout of `0`, if not 0, maybe `100` or so to delay the work in those so the UI/modals can present without any delay (hopefully)

**Comment by bradmartin** [comment]
E2 test data:

- first run: `4794ms`
- second run: `900ms`

Fairly consistent across app starts. Currently digging into native side profiling to see if can determine the slowdown, the logs pause for a moment at a specific place in the modal dialog section of the core modules for NS. Have looked at all the modal dialog pieces and currently don't see any "reason" for this to happen, so just want to rule out anything when profiling in android studio. Next, will look at the actual content of the modal being created to see if it's in relation to View count, etc.

**Comment by finger563** [comment]
Yea that matches up with about the timing I'm seeing (anecdotally) - glad to see the perf logs show it; it *shouldn't* be from the view complexity because it happens for all the modals - some of which (like the change settings, scanning, and about pages) are very simple

**Comment by bradmartin** [comment]
yea I didn't expect the Views to be the solution, I'll have more data on it shortly to see what's exactly going on

**Comment by bradmartin** [comment]
Current MainPage framestats:

`  103 views, 89.66 kB of display lists`

Settings modal UI:

`  32 views, 27.79 kB of display lists`

**Comment by bradmartin** [comment]
Have the E2 down to `1939ms` to open the settings modal vs. the `4794` time previously recorded prior to changes. Still tracking down the cause, for now it's just been a few minor tweaks to the modal code, none that I'd expect to actually improve the performance this much but.... it does 🤷‍♂

**Comment by bradmartin** [comment]
Just clocked a `964ms` on the E2. I'm gonna push up this branch changes, with the perf logs in place so anyone else can run it and test on real hardware while I dig deeper into the settings of the options to see what is the actual cause.

**Comment by bradmartin** [comment]
branch `bug/749` has code linked to tracking this down. Final fix not in place, need to clean it up before PR'ing and also determine WHY the commented section in `view.android.js` is causing the big slow down during modal creation.

**Comment by bradmartin** [comment]
Taking a short break but https://github.com/Max-Mobility/permobil-client/commit/9290717116a0e0f5e018d531496a25f45fdfd3f3#diff-8e8879c1d65dd8433bc00aaa4d7148b7R148 should show where the current events are slowing down and what is being analyzed further to find better solution to improve things without removing functionality.

**Comment by finger563** [comment]
awesome - thanks for tracking this down; I'm pulling now and building so I'll report the numbers here that I see on my watch.

**Comment by bradmartin** [comment]
I've been doing `onShownModally - settings modal open` from the `performance.now` calls. You'll see them when you open the `settings `modal in the terminal. Should give you a `xxxx.` number and i think it's correct in terms of it being `ms`.

**Comment by finger563** [comment]
Yea mine got `2.596 s` between `settings onShownModally` and `settings modal open`

**Comment by bradmartin** [comment]
ok so not greeeat, but better than the 4-6s I had been seeing. I'm looking into those options being passed and WHY some of those slow it down. It seems considerable for a couple calls to be seconds long in execution time

**Comment by bradmartin** [comment]
wait a second, did you pull the `onShownModally` from the 2nd modal in the `settings` opening for the translation download?

**Comment by finger563** [comment]
No - i'm doing the settings, not the scanning.  when testing the scanning page (e.g. when pressing `Pair SmartDrive`) I get ~ 2 seconds from this log:

```
JS: _showNativeModalView: 156550929.489737
JS: initializeDialogFragment: 156550933.177237
JS: initializeDialogFragment.DialogImpl: 156550943.828278
JS: initializeDialogFragment.DialogImpl: 156550945.062133
JS: initializeDialogFragment: 156550971.435049
JS: DialogFragment: 156550971.925466
JS: DialogFragment: 156551012.880987
JS: Bundle: 156551013.300362
JS: Bundle: 156551015.265153
JS: performanceNow: 156551016.939581
JS: performanceNow: 156551017.266039
JS: show: 156551017.960674
JS: show: 156551036.259424
JS: onCreateDialog: 156551324.038903
JS: onCreateDialog ownerId:97
JS: getModalOptions: 156551325.594164
JS: getModalOptions: 156551325.813695
JS: setting the options: 156551325.988174
JS: setting the _dialogFragment: 156551326.290622
JS: setting the _dialogFragment: 156551326.57281
JS: setStyle: 156551326.866143
JS: setStyle: 156551327.267758
JS: getting theme for dialog: 156551327.43807
JS: getting theme for dialog: 156551327.765674
JS: setting the options: 156551327.969737
JS: new DialogImpl: 156551328.196508
JS: new DialogImpl: 156551693.691247
JS: dialog android.app.Dialog_bundle_109849_26_DialogImpl@6ba9d80
JS: onCreateDialog: 156551695.285205
JS: DialogFragmentImpl.onCreateView: 156551697.255726
JS: DialogFragmentImpl.onCreateView: 156551912.720934
JS: DialogFragmentImpl.onStart: 156552137.13182
JS: DialogFragmentImpl._shownCallback: 156552287.427497
JS: DEBUG:  scanning-page onShownModally
```

**Comment by finger563** [comment]
on a re-run of the app (e.g. not after a clean but after a re-deploy) I get this log for the settings page to show (~1.2 seconds):
```
JS: settings modal open: 156664113.279277
JS: _showNativeModalView: 156664242.606777
JS: initializeDialogFragment: 156664243.157819
JS: initializeDialogFragment: 156664243.375475
JS: DialogFragment: 156664243.566881
JS: DialogFragment: 156664301.149694
JS: Bundle: 156664301.51636
JS: Bundle: 156664306.540266
JS: performanceNow: 156664308.481621
JS: performanceNow: 156664308.739069
JS: show: 156664309.206204
JS: show: 156664319.522819
JS: settings modal open: 156664319.95636
JS: onCreateDialog: 156664350.304641
JS: onCreateDialog ownerId:104
JS: getModalOptions: 156664351.320891
JS: getModalOptions: 156664351.713496
JS: setting the options: 156664351.941412
JS: setting the _dialogFragment: 156664352.246204
JS: setting the _dialogFragment: 156664352.471725
JS: setStyle: 156664352.745944
JS: setStyle: 156664353.224954
JS: getting theme for dialog: 156664353.447141
JS: getting theme for dialog: 156664353.7336
JS: setting the options: 156664353.916725
JS: new DialogImpl: 156664354.119589
JS: new DialogImpl: 156664482.77235
JS: dialog android.app.Dialog_bundle_109849_26_DialogImpl@c49caca
JS: onCreateDialog: 156664484.009433
JS: DialogFragmentImpl.onCreateView: 156664487.453235
JS: DialogFragmentImpl.onCreateView: 156664866.492298
JS: DialogFragmentImpl.onStart: 156665107.048912
JS: DialogFragmentImpl._shownCallback: 156665334.44386
JS: settings onShownModally: 156665334.978495
JS: DEBUG:  settings onShownModally
```

**Comment by bradmartin** [comment]
Yea, I'm def. seeing between `1-2s` now, if I remove the changes on `view.android.js` regarding modals and re-run the app it's been up to `6-8s`

**Comment by finger563** [comment]
Well at least there's an improvement - just gotta figure out a way to get it as fast as possible :)

**Comment by bradmartin** [comment]
yep... still digging, should be able to find something, seems odd but right now the `git blame` is your `if/else on the cancelable option` 😆

**Comment by finger563** [comment]
haha of course it is :P

**Comment by bradmartin** [comment]
IDK why tho 😁 just commenting this block:

```
  // if (this._cancelable) {
      //   // for swipe dismiss modals on WearOS apps
      //   theme = com.permobil.smartdrive.wearos.R.style.SwipeableActivityTheme;
      // }

      // if (this._fullscreen) {
      //   theme = this.getActivity().getApplicationInfo().theme;
      // }
``` 
it's usually `1-2s`, add it back and re-run the app and `6.8s+` we'll figure something out.

**Comment by bradmartin** [comment]
```
        theme = com.permobil.smartdrive.wearos.R.style.SwipeableActivityTheme;
```
took `3788ms`  😕

**Comment by finger563** [comment]
wow - that's crazy.  worst case we can remove the swipe-ability if it kills the performance that much :/

**Comment by bradmartin** [comment]
yea... makes no sense to me why that call is 3800ms on average

**Comment by finger563** [comment]
maybe it's reloading it every time, so if we cached it then it might be faster?

**Comment by bradmartin** [comment]
Alright @finger563 - tomorrow, pull latest on that branch and test for me and let me know what you see. Just moved the theme to a top level `var` and then using that `var` at modal execution time and it seems to be faster than calling it directly 👍 - still think we can gain some perf somewhere in there even more.

**Comment by finger563** [comment]
testing now :crossed_fingers:

**Comment by finger563** [comment]
closed by #757 and superseded by #758
